components/req_cpu/src/linux.rs

```diff
@@ -0,0 +1,690 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::collector::{Collector, CollectorId};
+use crate::{RecorderConfig, RequestCpuRecords, RequestTag};
+
+use std::cell::Cell;
+use std::fs::read_dir;
+use std::marker::PhantomData;
+use std::sync::atomic::Ordering::{AcqRel, Acquire, Relaxed};
+use std::sync::atomic::{AtomicPtr, AtomicU64};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use collections::{HashMap, HashSet};
+use crossbeam::channel::{unbounded, Receiver, Sender};
+use lazy_static::lazy_static;
+use libc::pid_t;
+use procinfo::pid;
+use procinfo::pid::Stat;
+
+pub fn init_recorder() {
+    lazy_static! {
+        static ref PHANTOM: () = {
+            std::thread::Builder::new()
+                .name("req-cpu-recorder".to_owned())
+                .spawn(move || {
+                    let mut recorder = ReqCpuRecorder::new(RecorderConfig::default());
+
+                    loop {
+                        recorder.handle_collector_registration();
+                        recorder.handle_thread_registration();
+                        recorder.record();
+                        recorder.may_advance_window();
+                        recorder.may_gc();
+
+                        std::thread::sleep(Duration::from_micros(
+                            (recorder.config.record_interval_ms * 1_000.0) as _,
+                        ));
+                    }
+                })
+                .expect("Failed to create recorder thread");
+        };
+    }
+    *PHANTOM
+}
+
+pub fn register_collector(collector: Box<dyn Collector>) -> CollectorHandle {
+    lazy_static! {
+        static ref NEXT_COLLECTOR_ID: Arc<AtomicU64> = Arc::new(AtomicU64::new(1));
+    }
+
+    let id = CollectorId(NEXT_COLLECTOR_ID.fetch_add(1, Relaxed));
+    COLLECTOR_REGISTRATION_CHANNEL
+        .0
+        .send(CollectorRegistrationMsg::Register { collector, id })
+        .ok();
+    CollectorHandle { id }
+}
+
+impl RequestTag {
+    pub fn attach(self: &Arc<Self>) -> Guard {
+        CURRENT_REQ.with(|s| {
+            if s.is_set.get() {
+                panic!("Nested attachment is not allowed.")
```

Comment:
In what case will this panic happen?

Reply:
For now, it can not happen. Tagged tasks are executed one by one; no nested tasks exist.